File: app/src/main/java/com/example/maunorafiq/pawangcuaca/presentation/forecast/ForecastPresenter.java
package com.example.maunorafiq.pawangcuaca.presentation.forecast;

import android.support.annotation.NonNull;

import com.example.maunorafiq.pawangcuaca.presentation.internal.di.PerActivity;
import com.example.maunorafiq.pawangcuaca.presentation.mapper.ForecastModelDataMapper;
import com.example.maunorafiq.pawangcuaca.presentation.mapper.WeatherModelDataMapper;
import com.example.maunorafiq.pawangcuaca.presentation.model.ForecastModel;
import com.example.maunorafiq.pawangcuaca.presentation.model.WeatherModel;
import com.example.maunorafiq.pawangcuaca.presentation.base.Presenter;
import com.icehousecorp.maunorafiq.domain.forecast.Forecast;
import com.icehousecorp.maunorafiq.domain.forecast.interactor.GetForecast;
import com.icehousecorp.maunorafiq.domain.weather.Weather;
import com.icehousecorp.maunorafiq.domain.weather.interactor.GetWeather;

import javax.inject.Inject;

import rx.Subscriber;

/**
 * Created by maunorafiq on 11/29/16.
 */
@PerActivity
public class ForecastPresenter implements Presenter {

    private final String TAG = this.getClass().getSimpleName();

    private ForecastView forecastView;

    private final GetForecast getForecastUseCase;
    private final GetWeather getWeatherUseCase;
    private final WeatherModelDataMapper weatherModelDataMapper;
    private final ForecastModelDataMapper forecastModelDataMapper;

    @Inject
    public ForecastPresenter(GetForecast getForecastUseCase,
                             GetWeather getWeatherUseCase,
                             WeatherModelDataMapper weatherModelDataMapper,
                             ForecastModelDataMapper forecastModelDataMapper) {
        this.getForecastUseCase = getForecastUseCase;
        this.getWeatherUseCase = getWeatherUseCase;
        this.weatherModelDataMapper = weatherModelDataMapper;
        this.forecastModelDataMapper = forecastModelDataMapper;
    }

    public void setView(@NonNull ForecastView forecastView) {
        this.forecastView = forecastView;
    }

    @Override
    public void resume() { }

    @Override
    public void pause() { }

    @Override
    public void destroy() {
        getForecastUseCase.unsubscribe();
        getWeatherUseCase.unsubscribe();
        forecastView = null;
    }

    public void initialize(String city) {
        loadForecast(city);
    }

    private void loadForecast(String city) {
        showViewLoading();
        hideViewRetry();
        getForecast(city);
    }

    private void showViewLoading() {
        forecastView.showLoading();
    }

    private void hideViewLoading() {
        forecastView.hideLoading();
    }

    private void showViewRetry() {
        forecastView.showRetry();
    }

    private void hideViewRetry() {
        forecastView.hideRetry();
    }

    private void showViewError(String message) {
        forecastView.showError(message);
    }

    private void showForecastInView(ForecastModel forecastModel) {
        forecastView.renderForecastWeather(forecastModel);
    }

    private void showForecastInView(WeatherModel weatherModel) {
        forecastView.renderForecastWeather(weatherModel);
    }

    private void getForecast(String city) {
        getForecastUseCase.setCity(city);
        getForecastUseCase.execute(new ForecastSubscriber());

        getWeatherUseCase.setCity(city);
        getWeatherUseCase.execute(new WeatherSubscriber());
    }

    private final class ForecastSubscriber extends Subscriber<Forecast> {
        @Override
        public void onCompleted() {
            hideViewLoading();
        }

        @Override
        public void onError(Throwable e) {
            showViewError(e.getMessage());
            hideViewLoading();
            showViewRetry();
        }

        @Override
        public void onNext(Forecast forecast) {
            showForecastInView(forecastModelDataMapper.transform(forecast));
        }
    }

    private final class WeatherSubscriber extends Subscriber<Weather> {
        @Override
        public void onCompleted() {
            hideViewLoading();
        }

        @Override
        public void onError(Throwable e) {
            showViewError(e.getMessage());
            hideViewLoading();
            showViewRetry();
        }

        @Override
        public void onNext(Weather weather) {
            showForecastInView(weatherModelDataMapper.transform(weather));
        }
    }
}

File: app/src/main/java/com/example/maunorafiq/pawangcuaca/presentation/forecast/ForecastPresenter.kt
package com.example.maunorafiq.pawangcuaca.presentation.forecast

import android.util.Log
import com.example.maunorafiq.pawangcuaca.presentation.base.Presenter

import com.example.maunorafiq.pawangcuaca.presentation.internal.di.PerActivity
import com.example.maunorafiq.pawangcuaca.presentation.mapper.ForecastModelDataMapper
import com.example.maunorafiq.pawangcuaca.presentation.mapper.WeatherModelDataMapper
import com.example.maunorafiq.pawangcuaca.presentation.model.ForecastModel
import com.example.maunorafiq.pawangcuaca.presentation.model.WeatherModel
import com.example.maunorafiq.pawangcuaca.presentation.forecast.ForecastView
import com.icehousecorp.maunorafiq.domain.forecast.Forecast
import com.icehousecorp.maunorafiq.domain.forecast.interactor.GetForecast
import com.icehousecorp.maunorafiq.domain.weather.Weather
import com.icehousecorp.maunorafiq.domain.weather.interactor.GetWeather

import javax.inject.Inject

import rx.Subscriber

/**
 * Created by maunorafiq on 11/29/16.
 */
@PerActivity
class ForecastPresenter
@Inject
constructor(private val getForecastUseCase: GetForecast,
            private val getWeatherUseCase: GetWeather,
            private val weatherModelDataMapper: WeatherModelDataMapper,
            private val forecastModelDataMapper: ForecastModelDataMapper) : Presenter {

    private val TAG = this.javaClass.simpleName

    private var forecastView: ForecastView? = null

    fun setView(forecastView: ForecastView) {
        this.forecastView = forecastView
    }

    override fun resume() {
    }

    override fun pause() {
    }

    override fun destroy() {
        getForecastUseCase.unsubscribe()
        getWeatherUseCase.unsubscribe()
        forecastView = null
    }

    fun initialize(city: String) {
        loadForecast(city)
    }

    private fun loadForecast(city: String) {
        showViewLoading()
        hideViewRetry()
        getForecast(city)
    }

    private fun showViewLoading() {
        forecastView!!.showLoading()
    }

    private fun hideViewLoading() {
        forecastView!!.hideLoading()
    }

    private fun showViewRetry() {
        forecastView!!.showRetry()
    }

    private fun hideViewRetry() {
        forecastView!!.hideRetry()
    }

    private fun showViewError(message: String?) {
        forecastView!!.showError(message)
    }

    private fun showForecastInView(forecastModel: ForecastModel) {
        forecastView!!.renderForecastWeather(forecastModel)
    }

    private fun showForecastInView(weatherModel: WeatherModel) {
        forecastView!!.renderForecastWeather(weatherModel)
    }

    private fun getForecast(city: String) {
        getForecastUseCase.setCity(city)
        getForecastUseCase.execute(ForecastSubscriber())

        getWeatherUseCase.setCity(city)
        getWeatherUseCase.execute(WeatherSubscriber())
    }

    private inner class ForecastSubscriber : Subscriber<Forecast>() {
        override fun onCompleted() {
            hideViewLoading()
        }

        override fun onError(e: Throwable) {
            showViewError(e.message)
            hideViewLoading()
            showViewRetry()
        }

        override fun onNext(forecast: Forecast) {
            showForecastInView(forecastModelDataMapper.transform(forecast))
        }
    }

    private inner class WeatherSubscriber : Subscriber<Weather>() {
        override fun onCompleted() {
            hideViewLoading()
        }

        override fun onError(e: Throwable) {
            showViewError(e.message)
            hideViewLoading()
            showViewRetry()
        }

        override fun onNext(weather: Weather) {
            showForecastInView(weatherModelDataMapper.transform(weather))
        }
    }
}
